platforms/core-runtime/serialization/src/main/java/org/gradle/internal/serialize/kryo/StringDeduplicatingKryoBackedDecoder.java

```diff
@@ -175,31 +185,39 @@ public String readString() throws EOFException {
     @Override
     public String readNullableString() throws EOFException {
         try {
-            int idx = readInt();
-            if (idx == -1) {
-                return null;
-            }
-            if (strings == null) {
-                strings = new String[INITIAL_CAPACITY];
-            }
-            String string = null;
-            if (idx >= strings.length) {
-                String[] grow = new String[strings.length * 3 / 2];
-                System.arraycopy(strings, 0, grow, 0, strings.length);
-                strings = grow;
-            } else {
-                string = strings[idx];
+            int index = readStringIndex();
+            switch (index) {
+                case 0:
```

Comment:
🤔 One potential trick: ATM we're wasting two first elements of the array. We can avoid this and replace switch with a single if, though I'm not sure if it affects performance much:
1. Let's use 0 for new string and 1 for null
2. Have `null` in the initial array: `INITIAL_CAPACITY_MARKER = new String [] { null }`.
3. Now we can change the switch into
```java
if (index == 0) return readNewString();
return strings[index - 1];
```

But I guess this goes to microbenchmarking territory.

Reply:
Interesting idea! I went with a variation that avoids the subtraction on every read.

Reply:
On a second thought, using the two constants on both encoder and decoder makes the connection clearer and it doesn't seem to be a clear performance advantage either way.